File: magellan-java/src/main/java/com/silwek/tools/magellan/java/Magellan.java
package com.silwek.tools.magellan.java;

import android.app.Activity;
import android.content.Intent;

import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.navigation.NavController;
import androidx.navigation.NavDirections;
import androidx.navigation.fragment.NavHostFragment;

import kotlin.Unit;

public class Magellan {
    public static NavController findNavController(Fragment fragment) {
        if (fragment == null) //Cancel
            return null;
        return NavHostFragment.findNavController(fragment);
    }

    /**
     * @return 0 if no destination or null Fragment/NavController
     */
    public static int getCurrentDestinationId(@Nullable Fragment fragment) {
        if (fragment == null) return 0;
        NavController navController = findNavController(fragment);
        return getCurrentDestinationId(navController);
    }

    /**
     * @return 0 if no destination or null NavController
     */
    public static int getCurrentDestinationId(@Nullable NavController navController) {
        if (navController != null && navController.getCurrentDestination() != null)
            return navController.getCurrentDestination().getId();
        else return 0;
    }

    public static void navigateUp(Activity activity, Intent defaultIntent) {
        MagellanExtCompat.Companion.navigateUp(activity, defaultIntent);
    }

    public static void navigateUp(Fragment fragment, Intent defaultIntent) {
        MagellanExtCompat.Companion.navigateUp(fragment, defaultIntent);
    }

    public static boolean navigateBack(Fragment fragment) {
        return MagellanExtCompat.Companion.navigateBack(fragment);
    }

    public static void onBackPress(Fragment fragment, MagellanAction action) {
        MagellanExtCompat.Companion.onBackPress(fragment, () -> {
            action.doAction();
            return Unit.INSTANCE;
        });
    }

    public static void safeNavigate(NavController navController, NavDirections direction) {
        MagellanExtCompat.Companion.safeNavigate(navController, direction);
    }

    public static void safeNavigate(NavController navController, int destId) {
        MagellanExtCompat.Companion.safeNavigate(navController, destId);
    }

    public static void safeNavigateTo(Fragment fragment, int destId) {
        MagellanExtCompat.Companion.safeNavigateTo(fragment, destId);
    }

    public static void safeNavigateTo(Fragment fragment, NavDirections direction) {
        MagellanExtCompat.Companion.safeNavigateTo(fragment, direction);
    }

    public static void popBackStack(Fragment fragment) {
        MagellanExtCompat.Companion.popBackStack(fragment);
    }

    public interface MagellanAction {
        void doAction();
    }
}

File: magellan/src/main/java/com/silwek/tools/magellan/Magellan.kt
package com.silwek.tools.magellan

import android.app.Activity
import android.content.Intent
import androidx.fragment.app.Fragment
import androidx.navigation.NavController
import androidx.navigation.NavDirections

object Magellan {

    fun navigateUp(activity: Activity?, defaultIntent: Intent) {
        if (activity == null) //Cancel
            return
        activity.navigateUp(defaultIntent)
    }

    fun navigateUp(fragment: Fragment?, defaultIntent: Intent) {
        if (fragment == null) //Cancel
            return
        fragment.navigateUp(defaultIntent)
    }

    fun navigateBack(fragment: Fragment?): Boolean {
        if (fragment == null) //Cancel
            return false
        return fragment.navigateBack()
    }

    inline fun onBackPress(fragment: Fragment?, crossinline action: () -> Unit) {
        if (fragment == null) //Cancel
            return
        fragment.onBackPress(action)
    }

    fun safeNavigate(navController: NavController?, direction: NavDirections) {
        if (navController == null) //Cancel
            return
        navController.safeNavigate(direction)
    }

    fun safeNavigateTo(fragment: Fragment?, destId: Int) {
        if (fragment == null) //Cancel
            return
        fragment.safeNavigateTo(destId)
    }

    fun safeNavigateTo(fragment: Fragment?, direction: NavDirections) {
        if (fragment == null) //Cancel
            return
        fragment.safeNavigateTo(direction)
    }

    fun popBackStack(fragment: Fragment?) {
        if (fragment == null) //Cancel
            return
        fragment.popBackStack()
    }
}